metadata/src/main/java/org/apache/kafka/metadata/MetadataParser.java

```diff
@@ -0,0 +1,115 @@
+ */
+
+package org.apache.kafka.metadata;
+
+import org.apache.kafka.common.metadata.MetadataRecordType;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ByteBufferAccessor;
+import org.apache.kafka.common.protocol.ObjectSerializationCache;
+import org.apache.kafka.common.utils.ByteUtils;
+
+import java.nio.ByteBuffer;
+
+public class MetadataParser {
+    public static final int MAX_SERIALIZED_EVENT_SIZE = 32 * 1024 * 1024;
+
+    private static short unsignedIntToShort(int val, String what) {
+        if (val > Short.MAX_VALUE) {
+            throw new MetadataParseException("Value for " + what + " was too large.");
+        }
+        return (short) val;
+    }
+
+    /**
+     * Parse the given buffer.
+     *
+     * @param buffer    The buffer.  Its offsets will be modified.
+     *
+     * @return          The metadata message.
+     */
+    public static ApiMessage read(ByteBuffer buffer) {
+        short type;
+        try {
+            type = unsignedIntToShort(ByteUtils.readUnsignedVarint(buffer), "type");
+        } catch (Exception e) {
+            throw new MetadataParseException("Failed to read variable-length type " +
+                "number: " + e.getMessage());
+        }
+        short version;
+        try {
+            version = unsignedIntToShort(ByteUtils.readUnsignedVarint(buffer), "version");
+        } catch (Exception e) {
+            throw new MetadataParseException("Failed to read variable-length " +
+                "version number: " + e.getMessage());
+        }
+        MetadataRecordType recordType = MetadataRecordType.fromId(type);
+        ApiMessage message = recordType.newMetadataRecord();
+        try {
+            message.read(new ByteBufferAccessor(buffer), version);
+        } catch (Exception e) {
+            throw new MetadataParseException(recordType + "#parse failed: " +
+                e.getMessage());
+        }
+        if (buffer.hasRemaining()) {
+            throw new MetadataParseException("Found " + buffer.remaining() +
+                " bytes of garbage after " + recordType);
+        }
+        return message;
+    }
+
+    /**
+     * Find the size of an API message and set up its ObjectSerializationCache.
+     *
+     * @param message   The metadata message.
+     * @param version   The metadata message version.
+     * @param cache     The object serialization cache to use.
+     *
+     * @return          The size
+     */
+    public static int size(ApiMessage message,
+                           short version,
+                           ObjectSerializationCache cache) {
+        long messageSize = message.size(cache, version);
+        long totalSize = messageSize +
+            ByteUtils.sizeOfUnsignedVarint(message.apiKey()) +
+            ByteUtils.sizeOfUnsignedVarint(version);
+        if (totalSize > MAX_SERIALIZED_EVENT_SIZE) {
+            throw new RuntimeException("Event size would be " + totalSize + ", but the " +
```

Comment:
nit: we could define a more specific exception.

Reply:
I'm not sure if there would be value in that, since I wouldn't expect any place to catch this specifically.